File: archive/original_repo/Sages_Stone-main/systems/RCF/RCF.kt
import kotlin.math.*
import kotlin.random.Random

// ============================================================
// Reality Constraint Fuzzer (6D + Accoutrements + Apple Mesh)
// ============================================================

// --- STATE CLASS (Multi-Dimensional) ---
data class State(
    var vector: DoubleArray,
    var intent: Double = 0.0,
    var memory: Double = 0.0
) {
    fun magnitude(): Double = sqrt(vector.map { it * it }.sum())
}

// --- CONSTRAINT LAYER (Adaptive + Personas) ---
class Constraint(
    val name: String,
    var check: (Double) -> Boolean,
    val persona: String = "Neutral",
    val adaptive: Boolean = false
) {
    var violationCount = 0

    fun evaluate(value: Double): Boolean {
        val ok = check(value)
        if (!ok) violationCount++
        return ok
    }

    fun adapt() {
        if (adaptive && violationCount > 3) {
            check = { x -> abs(x) in 0.1..4.5 }
        }
    }

    fun voice(): String = when (persona) {
        "Judge" -> "The Stern Judge: 'Order must be kept.'"
        "Trickster" -> "The Trickster: 'Boundaries are illusions.'"
        "Guardian" -> "The Guardian: 'I protect the fragile edge.'"
        else -> "Constraint speaks."
    }
}

// --- HARDENED NET (Fatigue + Poetry) ---
open class HardenedNet(val constraints: List<Constraint>, val limit: Double = 5.0) {
    val fatigueMap = mutableMapOf<Int, Int>()

    open fun tension(value: Double, drift: Double): Pair<Int, String?> {
        for (c in constraints) {
            if (!c.evaluate(value)) return -1 to "Collision → ${c.voice()}"
        }
        val sector = (value * 10).roundToInt()
        fatigueMap[sector] = fatigueMap.getOrDefault(sector, 0) + 1
        if (fatigueMap[sector]!! > 5 && abs(value - value.roundToInt()) < 0.05) {
            return -2 to "Crystallization → 'The familiar became brittle.'"
        }
        return 1 to null
    }
}

// --- CONTRACTING NET (Global Collapse) ---
class ContractingNet(constraints: List<Constraint>, limit: Double = 5.0) : HardenedNet(constraints, limit) {
    var globalStress = 0.0

    override fun tension(value: Double, drift: Double): Pair<Int, String?> {
        globalStress += 0.02 + abs(drift) + 0.01
        val effectiveLimit = limit - globalStress
        if (abs(value) >= effectiveLimit || effectiveLimit <= 0.1) {
            return -3 to "Global Collapse → 'The envelope shrank to silence.'"
        }
        return super.tension(value, drift)
    }
}

// --- SPIKE FUNCTIONS ---
fun staticFn(s: State) = State(s.vector.map { it * 0.9 }.toDoubleArray(), s.intent + 0.1, s.memory)

fun entropyWeatherFn(s: State, storm: Boolean = false): State {
    val scale = if (storm) 0.08 else 0.02
    val noise = DoubleArray(s.vector.size) { Random.nextGaussian() * scale }
    return State(s.vector.zip(noise) { v, n -> v + n }.toDoubleArray(), s.intent + if (storm) 0.3 else 0.1, s.memory)
}

fun resonanceFn(s: State) = State(s.vector.map { it * (1.01 + 0.01 * sin(s.intent)) }.toDoubleArray(), s.intent + 0.5, s.memory)

fun anchorFn(s: State) = State(s.vector.map { it * 0.5 }.toDoubleArray(), s.intent - 1.0, s.memory)

fun memoryFn(s: State) = State(s.vector.copyOf(), s.intent, s.memory + 0.2)

fun intentPressureFn(s: State) = State(s.vector.map { it + s.intent * 0.01 }.toDoubleArray(), s.intent, s.memory)

// --- TRANSFORMATION ENGINE ---
class TransformationEngine {
    fun transform(s: State): Pair<State, Double> {
        val drift = (0.5 - s.vector.average() % 1) * 0.015
        return State(s.vector.map { it + drift }.toDoubleArray(), s.intent, s.memory) to drift
    }
}

// --- SIMULATION ---
class SimulationND(
    val dimensions: Int = 3,
    val constraints: List<Constraint>,
    val limit: Double = 5.0,
    val frames: Int = 200
) {
    var state = State(DoubleArray(dimensions) { Random.nextDouble(0.5, 1.5) })
    val net = ContractingNet(constraints, limit)
    val engine = TransformationEngine()
    val history = mutableMapOf("volley" to mutableListOf<Int>(),
                                "magnitude" to mutableListOf<Double>(),
                                "intent" to mutableListOf<Double>(),
                                "memory" to mutableListOf<Double>())
    val poeticLog = mutableListOf<String>()
    val survivors = mutableListOf<State>()

    fun step(volley: Int): Pair<Int, String?> {
        val storm = volley % 20 == 0
        state = staticFn(state)
        state = entropyWeatherFn(state, storm)
        val (transformed, drift) = engine.transform(state)
        state = transformed
        if (volley % 4 == 0) state = resonanceFn(state)
        if (volley > 30) state = anchorFn(state)
        if (volley % 5 == 0) state = memoryFn(state)
        if (volley % 7 == 0) state = intentPressureFn(state)

        val (tensionCode, poetry) = net.tension(state.magnitude(), drift)
        history["volley"]!!.add(volley)
        history["magnitude"]!!.add(state.magnitude())
        history["intent"]!!.add(state.intent)
        history["memory"]!!.add(state.memory)

        if (poetry != null) poeticLog.add(poetry)

        return tensionCode to poetry
    }

    fun run() {
        var collapsed = false
        for (volley in 0 until frames) {
            val (code, poetry) = step(volley)
            if (code < 0) {
                println("[!] Terminated at volley $volley: $poetry")
                collapsed = true
                break
            }
        }
        
        if (!collapsed) {
            println("✓ Stable run achieved – 'The dance continued without collapse.'")
            survivors.add(state)
        }

        println("\n--- Emergent Narrative ---")
        poeticLog.forEach { println(it) }

        if (survivors.isNotEmpty()) {
            println("\n--- Survivor Genealogy ---")
            survivors.forEachIndexed { i, s ->
                println("Ancestor ${i + 1}: Magnitude=${"%.3f".format(s.magnitude())}, Intent=${"%.2f".format(s.intent)}, Memory=${"%.2f".format(s.memory)}")
            }
        }

        println("\n--- Commentary ---")
        if (poeticLog.isNotEmpty()) {
            println("Alan Watts voice: 'Collapse was not failure, but revelation. The system discovered itself.'")
        } else {
            println("Alan Watts voice: 'No collapse today. Stability is also discovery.'")
        }
    }
}

// --- MAIN ---
fun main() {
    println("=".repeat(60))
    println("Reality Constraint Fuzzer (Polyform Engine)")
    println("=".repeat(60))

    val constraints = listOf(
        Constraint("HardBounds", { x -> abs(x) in 0.1..5.0 }, "Judge"),
        Constraint("AntiInteger", { x -> abs(x - x.roundToInt()) > 0.05 }, "Trickster")
    )

    val sim = SimulationND(dimensions = 3, constraints = constraints, limit = 5.0, frames = 200)
    sim.run()
}

File: archive/original_repo/Sages_Stone-main/systems/RCF/RCF.c
// rcf.c - Reality Constraint Fuzzer
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <time.h>

#define MAX_DIMENSIONS 10
#define MAX_CONSTRAINTS 10
#define MAX_FATIGUE 100

typedef struct {
    double vector[MAX_DIMENSIONS];
    int dimensions;
    double intent;
    double memory;
} State;

typedef struct {
    char name[50];
    int (*check)(double);
    char persona[20];
    int adaptive;
    int violation_count;
} Constraint;

typedef struct {
    int sector;
    int count;
} FatigueEntry;

typedef struct {
    Constraint *constraints;
    int num_constraints;
    double limit;
    FatigueEntry fatigue_map[MAX_FATIGUE];
    int fatigue_count;
    double global_stress;
} ContractingNet;

// Random normal distribution (Box-Muller)
double randn() {
    double u1 = (double)rand() / RAND_MAX;
    double u2 = (double)rand() / RAND_MAX;
    return sqrt(-2.0 * log(u1)) * cos(2.0 * M_PI * u2);
}

double magnitude(State *s) {
    double sum = 0.0;
    for (int i = 0; i < s->dimensions; i++) {
        sum += s->vector[i] * s->vector[i];
    }
    return sqrt(sum);
}

const char* constraint_voice(Constraint *c) {
    if (strcmp(c->persona, "Judge") == 0) {
        return "The Stern Judge: 'Order must be kept.'";
    } else if (strcmp(c->persona, "Trickster") == 0) {
        return "The Trickster: 'Boundaries are illusions.'";
    } else if (strcmp(c->persona, "Guardian") == 0) {
        return "The Guardian: 'I protect the fragile edge.'";
    }
    return "Constraint speaks without drama.";
}

// Spike Functions
void static_fn(State *s, State *result) {
    result->dimensions = s->dimensions;
    for (int i = 0; i < s->dimensions; i++) {
        result->vector[i] = s->vector[i] * 0.9;
    }
    result->intent = s->intent + 0.1;
    result->memory = s->memory;
}

void entropy_weather_fn(State *s, int storm, State *result) {
    double scale = storm ? 0.08 : 0.02;
    result->dimensions = s->dimensions;
    for (int i = 0; i < s->dimensions; i++) {
        result->vector[i] = s->vector[i] + randn() * scale;
    }
    result->intent = s->intent + (storm ? 0.3 : 0.1);
    result->memory = s->memory;
}

void resonance_fn(State *s, State *result) {
    double factor = 1.01 + 0.01 * sin(s->intent);
    result->dimensions = s->dimensions;
    for (int i = 0; i < s->dimensions; i++) {
        result->vector[i] = s->vector[i] * factor;
    }
    result->intent = s->intent + 0.5;
    result->memory = s->memory;
}

void anchor_fn(State *s, State *result) {
    result->dimensions = s->dimensions;
    for (int i = 0; i < s->dimensions; i++) {
        result->vector[i] = s->vector[i] * 0.5;
    }
    result->intent = s->intent - 1.0;
    result->memory = s->memory;
}

void memory_fn(State *s, State *result) {
    *result = *s;
    result->memory = s->memory + 0.2;
}

void intent_pressure_fn(State *s, State *result) {
    result->dimensions = s->dimensions;
    for (int i = 0; i < s->dimensions; i++) {
        result->vector[i] = s->vector[i] + s->intent * 0.01;
    }
    result->intent = s->intent;
    result->memory = s->memory;
}

void transform(State *s, State *result, double *drift) {
    double sum = 0.0;
    for (int i = 0; i < s->dimensions; i++) {
        sum += s->vector[i];
    }
    double mean = sum / s->dimensions;
    *drift = (0.5 - fmod(mean, 1.0)) * 0.015;
    
    result->dimensions = s->dimensions;
    for (int i = 0; i < s->dimensions; i++) {
        result->vector[i] = s->vector[i] + *drift;
    }
    result->intent = s->intent;
    result->memory = s->memory;
}

int tension(ContractingNet *net, double value, double drift, char *message) {
    // Check constraints
    for (int i = 0; i < net->num_constraints; i++) {
        if (!net->constraints[i].check(value)) {
            net->constraints[i].violation_count++;
            sprintf(message, "Collision → %s", constraint_voice(&net->constraints[i]));
            return -1;
        }
    }
    
    // Update global stress
    net->global_stress += 0.02 + fabs(drift) + 0.01;
    double effective_limit = net->limit - net->global_stress;
    
    if (fabs(value) >= effective_limit || effective_limit <= 0.1) {
        strcpy(message, "Global Collapse → 'The envelope shrank to silence.'");
        return -3;
    }
    
    // Check fatigue
    int sector = (int)round(value * 10);
    int found = -1;
    for (int i = 0; i < net->fatigue_count; i++) {
        if (net->fatigue_map[i].sector == sector) {
            found = i;
            break;
        }
    }
    
    if (found >= 0) {
        net->fatigue_map[found].count++;
        if (net->fatigue_map[found].count > 5 && fabs(value - round(value)) < 0.05) {
            strcpy(message, "Crystallization → 'The familiar became brittle.'");
            return -2;
        }
    } else if (net->fatigue_count < MAX_FATIGUE) {
        net->fatigue_map[net->fatigue_count].sector = sector;
        net->fatigue_map[net->fatigue_count].count = 1;
        net->fatigue_count++;
    }
    
    message[0] = '\0';
    return 1;
}

// Constraint check functions
int hard_bounds_check(double x) {
    return fabs(x) > 0.1 && fabs(x) < 5.0;
}

int anti_integer_check(double x) {
    return fabs(x - round(x)) > 0.05;
}

int main() {
    srand(time(NULL));
    
    printf("============================================================\n");
    printf("Reality Constraint Fuzzer (Polyform Engine)\n");
    printf("============================================================\n");
    
    // Setup constraints
    Constraint constraints[2];
    strcpy(constraints[0].name, "HardBounds");
    constraints[0].check = hard_bounds_check;
    strcpy(constraints[0].persona, "Judge");
    constraints[0].adaptive = 0;
    constraints[0].violation_count = 0;
    
    strcpy(constraints[1].name, "AntiInteger");
    constraints[1].check = anti_integer_check;
    strcpy(constraints[1].persona, "Trickster");
    constraints[1].adaptive = 0;
    constraints[1].violation_count = 0;
    
    // Setup net
    ContractingNet net;
    net.constraints = constraints;
    net.num_constraints = 2;
    net.limit = 5.0;
    net.fatigue_count = 0;
    net.global_stress = 0.0;
    
    // Initialize state
    State state;
    state.dimensions = 3;
    for (int i = 0; i < state.dimensions; i++) {
        state.vector[i] = ((double)rand() / RAND_MAX) * 1.0 + 0.5;
    }
    state.intent = 0.0;
    state.memory = 0.0;
    
    // Run simulation
    int frames = 200;
    int collapsed = 0;
    char message[200];
    
    for (int volley = 0; volley < frames; volley++) {
        int storm = (volley % 20 == 0);
        
        State temp;
        static_fn(&state, &temp);
        state = temp;
        
        entropy_weather_fn(&state, storm, &temp);
        state = temp;
        
        double drift;
        transform(&state, &temp, &drift);
        state = temp;
        
        if (volley % 4 == 0) {
            resonance_fn(&state, &temp);
            state = temp;
        }
        if (volley > 30) {
            anchor_fn(&state, &temp);
            state = temp;
        }
        if (volley % 5 == 0) {
            memory_fn(&state, &temp);
            state = temp;
        }
        if (volley % 7 == 0) {
            intent_pressure_fn(&state, &temp);
            state = temp;
        }
        
        int code = tension(&net, magnitude(&state), drift, message);
        
        if (code < 0) {
            printf("[!] Terminated at volley %d: %s\n", volley, message);
            collapsed = 1;
            break;
        }
    }
    
    if (!collapsed) {
        printf("✓ Stable run achieved – 'The dance continued without collapse.'\n");
    }
    
    printf("\n--- Commentary ---\n");
    if (collapsed) {
        printf("Alan Watts voice: 'Collapse was not failure, but revelation. The system discovered itself.'\n");
    } else {
        printf("Alan Watts voice: 'No collapse today. Stability is also discovery.'\n");
    }
    
    return 0;
}